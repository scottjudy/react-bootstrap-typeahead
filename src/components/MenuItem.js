// @flow

import cx from 'classnames';
import React, { type Node } from 'react';

import { withItem } from '../behaviors/item';

import type { EventHandler } from '../types';

type MenuItemProps = {
  active?: boolean,
  children?: Node,
  className?: string,
  disabled?: boolean,
  href?: string,
  onClick?: EventHandler<HTMLElement>,
};

const BaseMenuItem = React.forwardRef<MenuItemProps, ?HTMLElement>((
  { active, children, className, disabled, onClick, ...props },
  ref
) => {
  return (
    <a
      {...props}
      className={cx('dropdown-item', { active, disabled }, className)}
      href={props.href || '#'}
      onClick={(e: SyntheticEvent<HTMLElement>) => {
        e.preventDefault();
        !disabled && onClick && onClick(e);
      }}
      ref={ref}>
      {children}
    </a>
  );
});

export { BaseMenuItem };
export default withItem(BaseMenuItem);
