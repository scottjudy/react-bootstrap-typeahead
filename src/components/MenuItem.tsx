import cx from 'classnames';
import React, { forwardRef, HTMLProps, MouseEvent } from 'react';

import { useItem, UseItemProps } from '../behaviors/item';

export interface MenuItemProps extends HTMLProps<HTMLAnchorElement> {
  active?: boolean;
  disabled?: boolean;
}

export const BaseMenuItem = forwardRef<HTMLAnchorElement, MenuItemProps>(
  ({ active, children, className, disabled, onClick, ...props }, ref) => {
    return (
      <a
        {...props}
        className={cx('dropdown-item', { active, disabled }, className)}
        href={props.href || '#'}
        onClick={(e: MouseEvent<HTMLAnchorElement>) => {
          e.preventDefault();
          !disabled && onClick && onClick(e);
        }}
        ref={ref}>
        {children}
      </a>
    );
  }
);

export default function MenuItem(props: UseItemProps<HTMLAnchorElement>) {
  return <BaseMenuItem {...useItem(props)} />;
}
