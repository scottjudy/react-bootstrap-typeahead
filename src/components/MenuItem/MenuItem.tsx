import cx from 'classnames';
import React, { forwardRef, HTMLAttributes, MouseEvent } from 'react';

import { useItem, UseItemProps } from '../../behaviors/item';

export interface BaseMenuItemProps extends HTMLAttributes<HTMLAnchorElement> {
  active?: boolean;
  disabled?: boolean;
  href?: string;
}

export const BaseMenuItem = forwardRef<HTMLAnchorElement, BaseMenuItemProps>(
  ({ active, children, className, disabled, onClick, ...props }, ref) => {
    return (
      <a
        {...props}
        className={cx('dropdown-item', { active, disabled }, className)}
        href={props.href || '#'}
        onClick={(e: MouseEvent<HTMLAnchorElement>) => {
          e.preventDefault();
          !disabled && onClick && onClick(e);
        }}
        ref={ref}>
        {children}
      </a>
    );
  }
);

export type MenuItemProps = UseItemProps<HTMLAnchorElement>;

export default function MenuItem(props: MenuItemProps) {
  return <BaseMenuItem {...useItem(props)} />;
}
