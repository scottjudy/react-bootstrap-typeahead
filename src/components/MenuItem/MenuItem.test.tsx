import React from 'react';

import * as baseStories from './BaseMenuItem.stories';
import * as stories from './MenuItem.stories';

import {
  composeStories,
  generateSnapshots,
  screen,
  render,
  userEvent,
} from '../../tests/helpers';

const { Default: BaseDefault, Active, Disabled } = composeStories(baseStories);
const { Default } = composeStories(stories);

describe('<BaseMenuItem>', () => {
  generateSnapshots(baseStories);

  it('triggers an event when clicked', () => {
    const onClick = jest.fn();
    render(<BaseDefault onClick={onClick} />);

    const item = screen.getByRole('link');
    userEvent.click(item);

    expect(onClick).toHaveBeenCalledTimes(1);
  });

  it('renders a default href', () => {
    render(<BaseDefault />);
    expect(screen.getByRole('link')).toHaveAttribute('href', '#');
  });

  it('renders an href passed to the MenuItem', () => {
    const href = '/some/link';
    render(<BaseDefault href={href} />);
    expect(screen.getByRole('link')).toHaveAttribute('href', href);
  });

  it('renders an active base menu item', () => {
    render(<Active />);
    expect(screen.getByRole('link')).toHaveClass('active');
  });

  it('renders a disabled base menu item', () => {
    const onClick = jest.fn();
    render(<Disabled disabled onClick={onClick} />);

    const item = screen.getByRole('link');
    userEvent.click(item);

    expect(item).toHaveClass('disabled');
    expect(onClick).toHaveBeenCalledTimes(0);
  });
});

describe('<MenuItem>', () => {
  generateSnapshots(stories);

  it('renders a MenuItem', () => {
    render(<Default />);

    const option = screen.getByRole('option');
    expect(option).toHaveAttribute('aria-label', 'test label');
    expect(option).toHaveAttribute('aria-selected', 'false');
    expect(option).toHaveAttribute('id', 'test-id-item-0');
  });

  it('conditionally calls `onInitialItemChange`', () => {
    const onInitialItemChange = jest.fn();
    const context = { onInitialItemChange };

    const { rerender } = render(
      <Default context={context} props={{ position: 1 }} />
    );
    expect(onInitialItemChange).toHaveBeenCalledTimes(0);

    rerender(<Default context={context} props={{ position: 0 }} />);
    expect(onInitialItemChange).toHaveBeenCalledTimes(1);
  });

  it('conditionally calls `onInitialItemChange`', () => {
    const onActiveItemChange = jest.fn();
    const context = {
      activeIndex: 1,
      onActiveItemChange,
    };

    const { rerender } = render(
      <Default context={context} props={{ position: 0 }} />
    );
    expect(onActiveItemChange).toHaveBeenCalledTimes(0);

    rerender(<Default context={context} props={{ position: 1 }} />);
    expect(onActiveItemChange).toHaveBeenCalledTimes(1);
  });

  it('changes the active state of the menu item', () => {
    render(<Default context={{ activeIndex: 0 }} />);

    const option = screen.getByRole('option');
    expect(option).toHaveClass('active');
    expect(option).toHaveAttribute('aria-selected', 'true');
  });

  it('sets the active state if it is the only result', () => {
    render(<Default context={{ isOnlyResult: true }} />);

    const option = screen.getByRole('option');
    expect(option).toHaveClass('active');
    expect(option).toHaveAttribute('aria-selected', 'true');
  });

  it('calls the click handlers', () => {
    const onClick = jest.fn();
    const onMenuItemClick = jest.fn();

    render(<Default context={{ onMenuItemClick }} props={{ onClick }} />);

    userEvent.click(screen.getByRole('option'));
    expect(onClick).toHaveBeenCalledTimes(1);
    expect(onMenuItemClick).toHaveBeenCalledTimes(1);
  });

  it('renders a disabled menu item', () => {
    const onClick = jest.fn();
    render(<Default props={{ disabled: true, onClick }} />);

    const item = screen.getByRole('option');
    userEvent.click(item);

    expect(item).toHaveClass('disabled');
    expect(onClick).toHaveBeenCalledTimes(0);
  });
});
