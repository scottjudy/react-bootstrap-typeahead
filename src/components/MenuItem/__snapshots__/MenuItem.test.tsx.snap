// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BaseMenuItem> Active story renders snapshot 1`] = `
<a
  className="dropdown-item active"
  href="#"
  onClick={[Function]}
>
  This is an active base menu item
</a>
`;

exports[`<BaseMenuItem> Default story renders snapshot 1`] = `
<a
  className="dropdown-item"
  href="#"
  onClick={[Function]}
>
  This is a base menu item
</a>
`;

exports[`<BaseMenuItem> Disabled story renders snapshot 1`] = `
<a
  className="dropdown-item disabled"
  href="#"
  onClick={[Function]}
>
  This is a disabled base menu item
</a>
`;

exports[`<MenuItem> Default story renders snapshot 1`] = `
<a
  aria-label="test label"
  aria-selected={false}
  className="dropdown-item"
  href="#"
  id="test-id-item-0"
  onClick={[Function]}
  onMouseDown={[Function]}
  role="option"
>
  This is a menu item
</a>
`;
